### PR TITLE
Add Weak.ptr_eq

### LINKED ISSUE #55981: Tracking issue for weak_ptr_eq
This is a tracking issue for `rc::Weak::ptr_eq` and `sync::Weak::ptr_eq`.

Feature gate: #![feature(weak_ptr_eq)].

Steps:

- [x] Implement the RFC (#55987).
- [ ] Stabilization PR.

---
Old issue:

`Rc` has a function called `ptr_eq` it compares if the `Rc` points to the value, not the values themselves. I would like to add the same function to `Weak`. I got an implementation but have two questions:

- Does this require an RFC? Or is this small enough to do without one.
- What to do with dangling pointer, e.g. created by `Weak::new()`. I would argue always return `false` if either one is dangling.

### PR DESCRIPTION
**Pull request by Thomasdezeeuw**
I hope the doc tests alone are good enough.

We also might want to discuss the dangling pointer case (from `Weak::new()`).

Updates #55981.

### CODE REVIEW

**Comment by sfackler** [comment]
There's no ABA problem here since the underlying Rc allocation isn't freed while there are outstanding Weaks, right?

I agree with the behavior for dangling Weaks.

We should define this for both Arc's and Rc's Weak.

**Comment by Thomasdezeeuw** [comment]
@sfackler the memory is only deallocated once the weak count hits 0 (starts at 1), see https://github.com/rust-lang/rust/blob/master/src/liballoc/rc.rs#L855-L857 and https://github.com/rust-lang/rust/blob/master/src/liballoc/rc.rs#L1297-L1301.

As for the dangling pointers; I don't think `Rc` or `Arc` allows for dangling pointers, they can only be created with an actual value, at least when using safe code. Only `Weak::new` will create a dangling pointer to avoid allocating.

**Comment by sfackler** [comment]
Ok cool, this approach should work then. Let's add the same method for Arc and this should be good to go.

**Comment by sfackler** [comment]
cc @rust-lang/libs

**Comment by Thomasdezeeuw** [comment]
Do you want the `Arc` and `sync::Weak` change in this PR? Or should I make another?

Also what should I do with the stable/unstable attribute?

**Comment by sfackler** [comment]
Let's add it in this PR.

The new functions should be added as unstable initially. We can repurpose #55981 as the tracking issue.

**Comment by sfackler** [comment]
oh, we should also add some basic tests.

**Comment by Thomasdezeeuw** [comment]
Anything I can do to progress this?

**Comment by sfackler** [comment]
Oh whoops - Github doesn't email on force pushes so I didn't see it was rebased :(

@bors r+

File: src/liballoc/sync.rs [hunk 1]
```diff
@@ -1131,6 +1131,55 @@ impl<T: ?Sized> Weak<T> {
...
+    ///
+    /// let third_rc = Arc::new(5);
+    /// let third = Arc::downgrade(&third_rc);
+    ///
+    /// assert!(!Weak::ptr_eq(&first, &third));
+    /// ```
+    ///
+    /// When using `Weak::new` the `Weak` is never equal to any other `Weak`
+    /// until it points to actual data (e.g. by setting it to a value created by
+    /// calling [`Arc::downgrade`]).
```
Comment (jethrogb) [inline src/liballoc/sync.rs]: I think this documentation is slightly confusion. Outside of `std`, I don't think there is any way to turn a reference obtained through `Weak::new` into a reference to something real. I think it should just say something like:

> References obtained through `Weak::new` are not equal to any other references, including other invocations of `Weak::new`.

(same applies to `rc::Weak`)

Comment (Thomasdezeeuw) [reply src/liballoc/sync.rs]: > Outside of std, I don't think there is any way to turn a reference obtained through Weak::new into a reference to something real.

I'm not sure what you mean by that. You can overwrite the value, e.g.

```rust
let mut weak_value = Weak::new();
let value = Rc::new(5);
weak_value = Rc::dowgrade(&value);
```

As for the comparing two `Weak::new()` with another I agree, but isn't the example enough? Currently no pointer created via `Weak::new()` is equal to any other pointer, including pointers also created via `Weak::new()`. That is what the documentations currently says.

Comment (jethrogb) [reply src/liballoc/sync.rs]: The way the documentation is written implies that you can take a `Weak::new` and somehow turn it into something that's not a dangling pointer. For example some `fn(self: &mut Weak<T>)`. Assignment and `mem::replace` don't count.

**Comment by stjepang** [comment]
I disagree with the handling of the dangling case.

Suppose we had an `AtomicWeak<T>` type and then wanted to implement a `get` method that loads it or initializes to 0 when dangling:

```rust
fn get(a: &AtomicWeak<i32>) -> Arc<i32> {
    loop {
        let weak = a.load();
        match weak.upgrade() {
            Some(arc) => return arc,
            None => {
                let arc = Arc::new(0);
                let prev = a.compare_and_set(&weak, Arc::downgrade(&arc));
                if Weak::ptr_eq(&prev, &weak) {
                    return arc;
                }
            }
        }
    }
}
```

With the current behavior of `Weak::ptr_eq` for dangling pointers, the CAS operation and the `if` check would not work. In fact, I think it would be impossible to implement this function as intended.

**Comment by sfackler** [comment]
So what should the handling be then? Are dangling weaks equal to each other? To all pointers?

**Comment by stjepang** [comment]
@sfackler Ah, I misunderstood by thinking "dangling weak" means "weak with strong_count = 0". :(

Never mind then, the current behavior seems good to me.

File: src/liballoc/sync.rs (hunk 1 above)
Comment (Thomasdezeeuw) [reply src/liballoc/sync.rs]: Sorry @jethrogb I still don't really see your point. You keep mentioning "dangling pointer", but the public documentation doesn't. Furthermore you can turn a variable with the current value of `Weak::new` into something else by using plain assignment. That is also possible in any function that takes `&mut Weak<T>`, so I don't get why that doesn't "count".

**Comment by Thomasdezeeuw** [comment]
@stjepang but that still doesn't make you example work, does it? I mean you can remove the if statement and upgrade on the next loop, but that does mean some more atomic instructions.

File: src/liballoc/sync.rs (hunk 1 above)
Comment (jethrogb) [reply src/liballoc/sync.rs]: "dangling pointer" and `Weak::new` are equivalent.

> Furthermore you can turn a variable with the current value of Weak::new into something else by using plain assignment.

Dropping the old value and writing something new in its place is *not* the same as changing the value.

Comment (jethrogb) [reply src/liballoc/sync.rs]: > @jethrogb "dangling pointer" and Weak::new are not equivalent. That is only how the current API is implementated. Previous versions of Rust (not sure what version changed it) actually allocated on the heap when calling Weak::new. A change was made to not allocate by default for most (collection) types, including Weak. However this happend without a change in the API. This means we could go back to that implementation and then Weak::new and "dangling pointer" aren't equivalent anymore.

Ok fine, I was using them equivalently, but I meant `Weak::new`. This point is completely orthogonal to my main point about the phrasing of the documentation.

Comment (jethrogb) [reply src/liballoc/sync.rs]: > When using `Weak::new` the `Weak` is never equal to any other `Weak` until it points to actual data

The "until it points to actual data" part clearly implies you can take a `Weak::new` and somehow change it to a value that is real. Dropping the old value and moving something new in its place is *not* the same as changing the value.

**Comment by stjepang** [comment]
@Thomasdezeeuw Hmmm, right... if `AtomicWeak` is initialized to `Weak::new()`, then it doesn't work.

@SimonSapin also has a good point.

I think there are two ways to interpret a dangling `Weak::new()` pointer:

1. It is like an uninitialized pointer and every such pointer is (potentially?) different.
2. It is like a null pointer and all such pointers are equal.

Given these two choices, I'd prefer to go with the second one and make all dangling pointers equal.

**Comment by stjepang** [comment]
Btw, a small nit: in the docs for [`std::rc::Weak::new`](https://github.com/rust-lang/rust/blob/90d2914da3d15ebe8ea0c95de2b3270ef3c8bddb/src/liballoc/rc.rs#L1191), the link to `upgrade` is dangling. Might be good to fix it in this PR.

**Comment by jethrogb** [comment]
I don't think they should be equal.

**Comment by stjepang** [comment]
@jethrogb But do you think a dangling pointer should at least be equal to itself?

E.g. `let w = Weak::new(); Weak::ptr_eq(&w, &w)`

**Comment by jethrogb** [comment]
In general I think `ptr_eq` should imply `eq`. Now I know `Weak` doesn't have `eq`, but I feel like that's too magic for similar reasons.

**Comment by Thomasdezeeuw** [comment]
> I don't think they should be equal.

> In general I think `ptr_eq` should imply `eq`.

@jethrogb, Isn't that contradicting? Also what is "too magic" about it, and for what reasons? Could you elaborate that point?

Also what do you think about @stjepang example?

**Comment by Thomasdezeeuw** [comment]
It's a week later, are we any closer to consensus whether or `Weak::ptr_eq(Weak::new(), Weak::new())` should return true or false?

**Comment by Thomasdezeeuw** [comment]
I've pushed changes to make `Weak::new()` equal itself. Can this be merged now? If people disagree with handling of `Weak::new()`, or any other cases, we have a stabilisation period for that right?

**Comment by sfackler** [comment]
Seems reasonable.

@bors r+

**Comment by chpio** [comment]
what is the reason for `ptr_eq` to be an associated function? I mean it's just a `Weak` not `Rc`, there's no deref?

**Comment by Thomasdezeeuw** [comment]
I just matched the method on `Rc`/`Arc`.

**Pull request merged**